// now/build/lib/plans.js
const ms = require('ms');

const Now = require('../lib');

function parsePlan(res) {return __async(function*(){
  let id;
  let until;

  const { subscription } = yield res.json();

  if (subscription) {
    id = subscription.plan.id;

    if (subscription.cancel_at_period_end) {
      until = ms(
        new Date(subscription.current_period_end * 1000) - new Date(),
        { long: true }
      );
    }
  } else {
    id = 'oss';
  }

  return { id, until };
}())}

module.exports = class Plans extends Now {
  getCurrent() {return __async(function*(){
    const res = yield this._fetch('/www/user/plan');
    return parsePlan(res);
  }.call(this))}

  set(plan) {return __async(function*(){
    const res = yield this._fetch('/www/user/plan', {
      method: 'PUT',
      body: { plan }
    });

    if (res.ok) {
      return parsePlan(res);
    }

    const err = new Error(res.statusText);
    err.res = res;
    throw err;
  }.call(this))}
};

function __async(g){return new Promise(function(s,j){function c(a,x){try{var r=g[x?"throw":"next"](a)}catch(e){j(e);return}r.done?s(r.value):Promise.resolve(r.value).then(c,d)}function d(e){c(e,1)}c()})}
